### ISSUE TITLE
Problem: bounty collection doesn't contain all the bounties

### ISSUE THREAD

**Issue by janat08** [Blockrazor/blockrazor#810]
Problem:  /bounties has to pull data from 3 other collections and then map them to bounty document, which is passable only because we're not approving bounties nor entirely completing them. Also I'm not sure about the reliance on methods that query different collections to tell how much the bounty is worth, nor "default" bounties such as add new coin that are available being identified by special id rather than a separate field.

Solution: adjust methods where needed to also map to bounty collection instead of mapping them on client side, or use peerDB which has an issue for giving examples in JS rather than CS (which don't entirely work).

**Comment by anbud** [comment]
I've built it like that mostly to prevent data redundancy. We are going to have a lot of currencies (probably in thousands), so mapping each currency to the Bounty collection as well would just waste space. I agree it would probably improve performance a little bit (although it's questionable), but there's no point in saving the data in two places at the same time. Pulling it from 3 collections also ensures that bounty data is fresh and up to date. Bounties that are in the database and that have a regular ID are the real thing, they are used to mark users progress through the bounty and they're deleted from the database when they're done. These other ones are just providing the metadata about bounties.

The main problem here is that I've built this bounty system on the foundation of the old bounty system (which is not currently in use), but I didn't plan on adding Problems and Currencies later on, as they were not planned yet. The structure is a little bit messy right now, but it's optimized, and refactoring it now is probably a low priority. The solution will definitely evolve even more.

Also, I haven't had any real experience with peerDB, how would it help here?

**Comment by janat08** [comment]
mongo doesn't care about space at all, and you keep data fresh by make sure that you also modify bounty collection which may get hairy which is where peerDB helps further improving separation of concerns.

**Comment by janat08** [comment]
the premise for mongo is that you copy all the data you need into a collection and only query that collection.

**Comment by gazhayes** [comment]
It's good that you logged this one so that we can have a discussion about it.

Different types of bounty are totally different, they *act* differently and they are *approved* differently. For example the bounties from /problem are completely different to adding a new coin. And who knows what new types of bounties we will add in a week, a month, a year. The way we do bounties could be totally different in a few months, no one knows.

I could be wrong, but if we are completely honest with each other, I have a feeling you are more worried about what the code looks like than an any performance benefits? 

### Reminder about how and why we use the C4:
We have no idea what Blockrazor will look like in a month. You could do hours and hours of work fixing this problem and then we throw it all away in a month because we do something different. 

That's why you want to be *really* sure that something really is a problem before you solve it (i.e. it is causing pain to other people), and then you want to solve it in the simplest way possible. The more work that is required to fix the problem, the more confident you should be that it really is a problem and that the problem annoys other people not just you. That way, your code has a very good chance of surviving anyway, and even if some of your code is thrown away in a month it will probably only be a small amount compared to the amount of work you have done in total. Anyway, by posting these issues, you are doing exactly the right thing, finding out if other people also think this is a problem worth solving right now or not.

Remember this **isn't** a corporate product with central planning and top-down control, there's no _vision_ of what the final product will look like (>80% of all code that is written globally every year is discarded within 2 years because the vision was a dumb idea and didn't actually solve anyone's problems). Great code that no one uses is shit code, and shit code that everyone uses is great code. Instead, Blockrazor itself is trying to discover and solve a *real* problem, and it can only do that by solving lots of small problems (that's why everything we do is problem:... solution:...).

**Comment by janat08** [comment]
No, as I said the whole point behind mongo is that it's designed and allows for embedded fields which are to contain copied data to avoid aggregation.

The data that pertains to whether bounty has been approved the part that will genuinely be queried by users would typically reside within a single collection, which does have the upside that you don't have to figure out how it should look within every single collection.

**Comment by gazhayes** [comment]
> The data that pertains to whether bounty has been approved the part that will genuinely be queried by users would typically reside within a single collection, which does have the upside that you don't have to figure out how it should look within every single collection were it not denormalized.

Can you rephrase that part, I don't think I understand what you mean.

**Comment by janat08** [comment]
I meant that the code that looks good is easy to work with. having the single collection with all the fields needed for your view like /bounties makes it readily apparent what you're working with. Neither the problems nor the bounties are optimized as you're doubling number of documents in problems collection which then be indexed and doubling the number of queries with lack of uniform collection to work with on that route requiring extra publication that then do diffing for all documents in all 3 collections and then you could perhaps try using a query that has a limit to it so that we don't load 3000 documents without using aggregation.

**Comment by gazhayes** [comment]
Ok thanks, that's much easier to understand.

I'll think a bit more carefully about this and reply later on :)

**Comment by gazhayes** [comment]
> I meant that the code that looks good is easy to work with.

Yes I agree, and we should always make it easy to work with whenever it's convenient. But there's no point in reworking code just to make it easier to work with, that makes sense in a centrally planned environment but not with the C4. And with this specific bounties case, I doubt it will ever be easy to work with, there's simply too much variance as I mentioned above (different bounties have almost nothing in common). We might not even have them on the same page in the UI for much longer anyway, e.g. developer bounties might move somewhere else or whatever.

Another thing to consider is that although there might only be 3000 - 5000 currencies, this can easily translate into 60,000 - 100,000 individual bounties very quickly. So no matter what you do now, it's very unlikely that we will be using a single mongo collection for very long (if Blockrazor actually works). We may not even use mongo at all, we could be on Redis, or we could be doing something totally different.

No one knows what will happen of course, but it it's unlikely that we would have less than 10,000 bounties, even in the best/worst case.

> doubling number of documents in problems collection

I haven't even looked at that part of the codebase, we are copying problems to bounties instead of using them directly?

> lack of uniform collection to work with on the /bounties route

Even if all the bounties were in a single collection, each type of bounty is still going to be treated very differently. They do not (and *cannot*) have a uniform way to be created, approved, rejected, etc.

**Comment by anbud** [comment]
@gazhayes, that's about what I wanted to say.
Even if we place them all in a single collection, it won't stay for long. I've tried doing that already when the bounty number was small. And no, problems are used directly from the `Problems` collection, they're not copied to the `Bounty` collection currently.

Also, I can't agree with @janat08 that MongoDB doesn't care about space. I've worked on some big data projects where we experimented with MongoDB/NodeJS combo, and trust me, as the data grows, queries get slower and slower (even with proper indexing), so it takes a performance hit.

**Comment by janat08** [comment]
@anbud it got slower because you denormolized, querying subset of documents across 3 collections or more would've been faster?

@gazhayes what we do now is sort documents on client because we have them all, and we can do that only because we have all of them. asking for 50 documents sorted when there're more then that can not happen on clientk, unless you know were willing to create another collection with an index that sorts them or use aggregation which publication I don't think support with packages.

Regarding Problems collection, the comments and topics are individual documents.
